### ISSUE TITLE
Schema changes slow on large databases

### ISSUE THREAD

**Issue by jpeyper** [zendesk/maxwell#224]
Continuing from the discussion here: https://github.com/zendesk/maxwell/issues/223

The performance issue we have is that we have a database with a lot of databases, tables and columns.

Once Maxwell has done its initial load (which takes about 30-45 minutes) it looks like this.

```
SELECT COUNT(*) FROM maxwell.databases;
count(*)  
----------
       282
```
```
SELECT COUNT(*) FROM maxwell.tables;
COUNT(*)  
----------
     22516
```
```
SELECT COUNT(*) FROM maxwell.columns;
COUNT(*)  
----------
    947392
```

in my config i have this
```
...
include_dbs=asp,bbs,bip,bss,global,globip,jrb,map,tspa
include_tables=contact,investor
...
```

The tables I'm wanting to get events for are actually very static, almost never changing in structure (maybe once a year).

There are lots of tables and databases and tables I don't care about, and likely never will.

But every time a table gets created or altered in another schema Maxwell stops for many minutes while it deals with the structure change, before continuing through the log.

As @osheroff rightfully pointed out, the issue isn't that Maxwell tracks the entire server's structure changes, its that its slow at this scale.

As I understand, the reason it tracks all structure changes is so if you change your config to monitor different tables it is ready to go.

I guess I'd like to discuss what options there are to solve this.

A few ideas I've had are:

- Improve the performance to a point where I don't care that its doing it (not sure if that's even possible).
- Maxwell only tracks structure changes on included schemas / tables, and on startup checks whether it needs to load any new information into its tables.
- Maxwell has a list of schemas / tables that it ignores for schema changes, and I accept the risk that if I ever do want to track them then I'll have to deal with resetting Maxwell.
- Deal with it
- something amazing that I'm not nearly clever enough to think of...

**Comment by osheroff** [comment]
> Improve the performance to a point where I don't care that its doing it (not sure if that's even possible).

I've begun some exploratory work on this, in which I keep only one full snapshot, and then maintain deltas to the snapshot.  It should make the cost of a schema update a single row insert, which is probably good enough.  

> Maxwell has a list of schemas / tables that it ignores for schema changes, and I accept the risk that if I ever do want to track them then I'll have to deal with resetting Maxwell.

sounds like @nmaquet will take a whack at this, see https://github.com/zendesk/maxwell/issues/252

**Comment by osheroff** [comment]
@jpeyper I just read this... maxwell takes 30-45 minutes to do its initial schema load?  that seems crazy slow...

**Comment by jpeyper** [comment]
Given those table counts, how long would you expect it to take? 

I can get some timings off the box tomorrow of thst helps? 

Also, my dba complained at me because one of our testing environments hard drive filled up, and the maxwell schema was using 150gb.

**Comment by jpeyper** [comment]
These are the logs from my testing environment starting maxwell and doing an initial schema load (its in reverse order because its coming out of splunk, not off the box itself

```
2016-02-03T14:48:48,017 INFO  snaq.db.ConnectionPool.ReplicationConnectionPool: ReplicationConnectionPool: Getting connection (user/password): jdbc:mysql://asisdevdb1.awmltd.com.au:54020?useCursorFetch=true&zeroDateTimeBehavior=convertToNull
2016-02-03T14:48:48,195 INFO  snaq.db.ConnectionPool.ReplicationConnectionPool: ReplicationConnectionPool: Created a new connection
2016-02-03T14:48:48,260 WARN  com.zendesk.maxwell.MaxwellMysqlStatus: Warning: binlog_row_image is set to MINIMAL.  This may not be what you want.
2016-02-03T14:48:48,260 WARN  com.zendesk.maxwell.MaxwellMysqlStatus: See http://maxwells-daemon.io/compat for more information.
2016-02-03T14:48:48,395 INFO  snaq.db.ConnectionPool.ReplicationConnectionPool: ReplicationConnectionPool: Getting connection (user/password): jdbc:mysql://asisdevdb1.awmltd.com.au:54020?useCursorFetch=true&zeroDateTimeBehavior=convertToNull
2016-02-03T14:48:48,565 INFO  snaq.db.ConnectionPool.ReplicationConnectionPool: ReplicationConnectionPool: Created a new connection
2016-02-03T14:48:48,898 INFO  com.zendesk.maxwell.Maxwell: Maxwell is capturing initial schema
2016-02-03T14:49:00,216 INFO  snaq.db.ConnectionPool.ReplicationConnectionPool: ReplicationConnectionPool: Destroyed connection
2016-02-03T15:40:45,909 INFO  org.apache.kafka.clients.producer.ProducerConfig: ProducerConfig values: 
        compression.type = gzip
        metric.reporters = []
        metadata.max.age.ms = 300000
        metadata.fetch.timeout.ms = 5000
        acks = 1
        batch.size = 16384
        reconnect.backoff.ms = 10
        bootstrap.servers = [kafka1.eventstream-devtest.hq.local:9092, kafka2.eventstream-devtest.hq.local:9092, kafka3.eventstream-devtest.hq.local:9092]
        receive.buffer.bytes = 32768
        retry.backoff.ms = 100
        buffer.memory = 33554432
        timeout.ms = 30000
        key.serializer = class org.apache.kafka.common.serialization.StringSerializer
        retries = 0
        max.request.size = 1048576
        block.on.buffer.full = true
        value.serializer = class org.apache.kafka.common.serialization.StringSerializer
        metrics.sample.window.ms = 30000
        send.buffer.bytes = 131072
        max.in.flight.requests.per.connection = 5
        metrics.num.samples = 2
        linger.ms = 0
        client.id = 

2016-02-03T15:40:46,030 WARN  org.apache.kafka.clients.producer.ProducerConfig: The configuration value.serializer = class org.apache.kafka.common.serialization.StringSerializer was supplied but isn't a known config.
2016-02-03T15:40:46,031 WARN  org.apache.kafka.clients.producer.ProducerConfig: The configuration key.serializer = class org.apache.kafka.common.serialization.StringSerializer was supplied but isn't a known config.
2016-02-03T15:41:05,200 INFO  com.google.code.or.net.impl.TransportImpl: connecting to host: asisdevdb1.awmltd.com.au, port: 54020
2016-02-03T15:41:05,261 INFO  com.google.code.or.net.impl.TransportImpl: connected to host: asisdevdb1.awmltd.com.au, port: 54020, context: AbstractTransport.Context[threadId=4539351,scramble==PZSp\Z2(q"PP"S\TqZQ,protocolVersion=10,serverHost=asisdevdb1.awmltd.com.au,serverPort=54020,serverStatus=2,serverCollation=8,serverVersion=5.6.14-enterprise-commercial-advanced-log,serverCapabilities=63487]
2016-02-03T15:41:05,262 INFO  com.google.code.or.net.impl.AuthenticatorImpl: start to login, user: asiscdc, host: asisdevdb1.awmltd.com.au, port: 54020
2016-02-03T15:41:05,284 INFO  com.google.code.or.net.impl.AuthenticatorImpl: login successfully, user: asiscdc, detail: OKPacket[packetMarker=0,affectedRows=0,insertId=0,serverStatus=2,warningCount=0,message=<null>]
```

as you can see, the initial load took 52 minutes to get to the next log message (ignoring the ReplicationConnectionPool stuff, not sure what that is but its awfully noisy).

**Comment by osheroff** [comment]
@jpeyper btw I found a very large slowdown with how maxwell captures schemas in AWS/aurora.  Maybe it's specific to a mysql version, I will have to test with 5.6.14...

**Comment by osheroff** [comment]
@jpeyper released v1.1.0-pre1 (actually I already released that verison, but oops, I released it again) with much faster schema processing.  Would love to know how it works for you.

**Comment by jpeyper** [comment]
Great, I'll check it out soon.

**Comment by osheroff** [comment]
fixed in the 1.1 release
